# Этап 2. Исследование и подготовка данных
1. Загрузите файл snsdata.csv в рабочее пространство Google Colaboratory.
2. Начните работу с кейсом, добавив импорт библиотеки pandas.
3. Прочитайте файл данных snsdata.csv, используя библиотеку pandas, получите сводную
информацию по фрейму данных data.
4. Изучите значения переменной gender, используя value_counts()
5. Если изучить другие переменные во фрейме данных, то обнаружится, что, кроме пола,
отсутствующие значения есть только у переменной возраста (age). Для числовых
данных можно получить статистическую сводку с помощью describe().
6. Измените значения возраста меньшие 13 и большие 20 на NAN.
7. Запросите статистику по переменной age.
8. Выполните фиктивное кодирование для переменных женского и неопределенного
пола.
9. Посмотрите, как изменилась структура фрейма данных data.
10. Чтобы убедиться, что все сделано правильно, сравните созданные фиктивные
переменные с исходной переменной gender.
11. Рассчитайте средний возраст школьников, используя mean модуля DataFrame
библиотеки pandas.
12. Вычислите среднее значение возраста для 2006, 2007, 2008, 2009 годов выпуска.
Средний возраст каждого года выпуска отличается примерно на один год. Это является
хорошим подтверждением того, что наши данные разумны.
13. Замените недостающие данные возраста в фрейме данных data на среднее значение
возраста для данного года выпуска.
14. Проверьте, остались ли пустые значения в переменной age.

In [2]:
# Импортируем необходимые библиотеки
import pandas as pd
import numpy as np

# Шаг 2.1 - 2.3
data = pd.read_csv('snsdata.csv')

data_info = data.info()
data_head = data.head()
data_describe = data.describe()

data_info, data_head, data_describe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 40 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   gradyear      30000 non-null  int64  
 1   gender        27276 non-null  object 
 2   age           24914 non-null  float64
 3   friends       30000 non-null  int64  
 4   basketball    30000 non-null  int64  
 5   football      30000 non-null  int64  
 6   soccer        30000 non-null  int64  
 7   softball      30000 non-null  int64  
 8   volleyball    30000 non-null  int64  
 9   swimming      30000 non-null  int64  
 10  cheerleading  30000 non-null  int64  
 11  baseball      30000 non-null  int64  
 12  tennis        30000 non-null  int64  
 13  sports        30000 non-null  int64  
 14  cute          30000 non-null  int64  
 15  sex           30000 non-null  int64  
 16  sexy          30000 non-null  int64  
 17  hot           30000 non-null  int64  
 18  kissed        30000 non-nu

(None,
    gradyear gender     age  friends  basketball  football  soccer  softball  \
 0      2006      M  18.982        7           0         0       0         0   
 1      2006      F  18.801        0           0         1       0         0   
 2      2006      M  18.335       69           0         1       0         0   
 3      2006      F  18.875        0           0         0       0         0   
 4      2006    NaN  18.995       10           0         0       0         0   
 
    volleyball  swimming  ...  blonde  mall  shopping  clothes  hollister  \
 0           0         0  ...       0     0         0        0          0   
 1           0         0  ...       0     1         0        0          0   
 2           0         0  ...       0     0         0        0          0   
 3           0         0  ...       0     0         0        0          0   
 4           0         0  ...       0     0         2        0          0   
 
    abercrombie  die  death  drunk  drugs  
 0 

In [4]:
# Шаг 2.4 gender и включая NaN
gender_distribution = data['gender'].value_counts(dropna=False)

# Шаг 2.5
age_describe_before = data['age'].describe()

# Шаг 2.6 Меняем возраст, выходящий за пределы 13-20 лет, на NaN
data.loc[(data['age'] < 13) | (data['age'] > 20), 'age'] = np.nan

# Шаг 2.7 Статистику по переменной age после очистки
age_describe_after = data['age'].describe()

print(f'Распределение по полу (включая NaN):\n{gender_distribution}')
print(f'\nСтатистика по возрасту до очистки:\n{age_describe_before}')
print(f'\nСтатистика по возрасту после очистки:\n{age_describe_after}')

Распределение по полу (включая NaN):
gender
F      22054
M       5222
NaN     2724
Name: count, dtype: int64

Статистика по возрасту до очистки:
count    24477.000000
mean        17.252429
std          1.157465
min         13.027000
25%         16.304000
50%         17.265000
75%         18.220000
max         19.995000
Name: age, dtype: float64

Статистика по возрасту после очистки:
count    24477.000000
mean        17.252429
std          1.157465
min         13.027000
25%         16.304000
50%         17.265000
75%         18.220000
max         19.995000
Name: age, dtype: float64


In [5]:
# Шаг 2.8 
data['no_gender'] = data['gender'].replace(['F', 'M', np.nan], [0, 0, 1])
data['female'] = data['gender'].replace(['F', 'M', np.nan], [1, 0, 0])

In [6]:
# Шаг 2.9 Структура фрейма данных после фиктивного кодирования
data_info_after_dummy = data.info()
data_info_after_dummy

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 42 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   gradyear      30000 non-null  int64  
 1   gender        27276 non-null  object 
 2   age           24477 non-null  float64
 3   friends       30000 non-null  int64  
 4   basketball    30000 non-null  int64  
 5   football      30000 non-null  int64  
 6   soccer        30000 non-null  int64  
 7   softball      30000 non-null  int64  
 8   volleyball    30000 non-null  int64  
 9   swimming      30000 non-null  int64  
 10  cheerleading  30000 non-null  int64  
 11  baseball      30000 non-null  int64  
 12  tennis        30000 non-null  int64  
 13  sports        30000 non-null  int64  
 14  cute          30000 non-null  int64  
 15  sex           30000 non-null  int64  
 16  sexy          30000 non-null  int64  
 17  hot           30000 non-null  int64  
 18  kissed        30000 non-nu

In [9]:
# Шаг 2.10 Сравниваем созданные фиктивные переменные с исходной переменной gender
gender_comparison = data[['gender', 'female', 'no_gender']].head()
gender_comparison

gender  female  no_gender
0      M     0.0        0.0
1      F     1.0        0.0
2      M     0.0        0.0
3      F     1.0        0.0
4    NaN     0.0        1.0

In [11]:
# Шаг 2.11 Средний возраст школьников
mean_age = data['age'].mean()
mean_age

17.252428933284307

In [12]:
# Шаг 2.12 Вычисляем средний возраст для каждого года выпуска (2006, 2007, 2008, 2009)
mean_age_by_gradyear = data[data['gradyear'].isin([2006, 2007, 2008, 2009])].groupby('gradyear')['age'].mean()
mean_age_by_gradyear

gradyear
2006    18.655858
2007    17.706172
2008    16.767701
2009    15.819573
Name: age, dtype: float64

In [14]:
# Шаг 2.13 Заменяем недостающие данные возраста на среднее значение возраста для данного года выпуска
for year in [2006, 2007, 2008, 2009]:
    avg_age_for_year = data.loc[data['gradyear'] == year, 'age'].mean()
    data.loc[(data['gradyear'] == year) & (data['age'].isna()), 'age'] = avg_age_for_year

# Шаг 2.14 Проверяем, остались ли пустые значения в переменной age
missing_age_count = data['age'].isna().sum()
missing_age_count

0

# Этап 3. Обучение модели на данных
Начните кластерный анализ с рассмотрения 36 признаков, показывающих, сколько раз
различные интересы появлялись в профилях социальной сети подростков.
1. Для удобства постройте фрейм данных, содержащий только эти признаки.
2. Выполните стандартизацию по z-оценке.
3. Сравните сводную статистику столбца basketball в старых и новых данных об интересах.
4. Выполните кластеризацию методом k-средних, с пятью кластерами.

In [20]:
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans


# # Шаг 3.1
interest_columns = [
    'basketball', 'football', 'soccer', 'softball', 'volleyball', 'swimming',
    'cheerleading', 'baseball', 'tennis', 'sports', 'cute', 'sex', 'sexy',
    'hot', 'kissed', 'dance', 'band', 'marching', 'music', 'rock', 'god',
    'church', 'jesus', 'bible', 'hair', 'dress', 'blonde', 'mall', 'shopping',
    'clothes', 'hollister', 'abercrombie', 'die', 'death', 'drunk', 'drugs'
]

interest_data = data[interest_columns]

# Шаг 3.2 Выполняем стандартизацию по z-оценке
# Вычисляет среднее значение признака и вычитает его из всех значений признака, делит на стандартное отклонение
scaler = StandardScaler()
interest_data_scaled = scaler.fit_transform(interest_data)

interest_data_scaled_df = pd.DataFrame(interest_data_scaled, columns=interest_columns)

# Шаг 3.3 Сравниваем сводную статистику столбца basketball в старых и новых данных
basketball_stats_before = interest_data['basketball'].describe()
basketball_stats_after = interest_data_scaled_df['basketball'].describe()

# Шаг 3.4 Выполняем кластеризацию методом k-средних с 5 кластерами
kmeans = KMeans(n_clusters=5, random_state=42, n_init='auto')
kmeans.fit(interest_data_scaled_df)
cluster_labels = kmeans.labels_

# Выводим результаты
print(f'Статистика по столбцу basketball до:\n{basketball_stats_before}')
print(f'\nСтатистика по столбцу basketball после:\n{basketball_stats_after}')
print(f'\nМетки кластеров для первых 10 наблюдений:\n{cluster_labels[:10]}')



Статистика по столбцу basketball до:
count    30000.000000
mean         0.267333
std          0.804708
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max         24.000000
Name: basketball, dtype: float64

Статистика по столбцу basketball после:
count    3.000000e+04
mean     5.494864e-17
std      1.000017e+00
min     -3.322173e-01
25%     -3.322173e-01
50%     -3.322173e-01
75%     -3.322173e-01
max      2.949277e+01
Name: basketball, dtype: float64

Метки кластеров для первых 10 наблюдений:
[0 1 0 0 1 0 4 0 0 0]


Статистика по столбцу basketball до стандартизации:
Среднее значение: 0.267 — в среднем, подростки редко упоминают баскетбол в своих профилях.
Минимум/медиана/максимум: Большинство подростков вообще не упоминают баскетбол (медиана — 0, 75% не упоминают баскетбол).
Максимум: 24 упоминания в одном профиле — некоторые пользователи явно сильно заинтересованы в этом виде спорта.
Стандартное отклонение: 0.804 — большая вариативность в данных, некоторые подростки могут упоминать баскетбол значительно чаще.

Статистика по столбцу basketball после стандартизации:
Среднее значение: 5.49e-17 — очень близко к нулю, что соответствует нормализации данных.
Минимум/медиана/максимум: Если значение меньше 0, это говорит о том, что баскетбол упоминается реже среднего, а если больше 0 — чаще.
Максимум: 29.49 — баскетбол может быть упомянут гораздо чаще в некоторых профилях по сравнению с остальными.
Стандартное отклонение: 1 — это стандарт для z-оценки (данные масштабируются так, чтобы отклонения от среднего были измеримы в единицах стандартного отклонения).

# Этап 4. Повышение эффективности модели
1. Для исследования кластеризации рассмотрите координаты центроидов кластеров
2. Выясните, к какому кластеру относятся первые десять школьников в таблице данных

In [24]:
# Шаг 4.1 Координаты центроидов кластеров
centroids = kmeans.cluster_centers_

# Преобразуем центроиды в DataFrame
centroids_df = pd.DataFrame(centroids, columns=interest_columns)

# Шаг 4.2 Выясняем, к какому кластеру относятся первые десять школьников
first_ten_students_clusters = cluster_labels[:10]

print(f'Центроиды кластеров:\n{centroids_df}')
print(f'\nКластеры первых десяти школьников:\n{first_ten_students_clusters}')


Центроиды кластеров:
   basketball  football    soccer  softball  volleyball  swimming  \
0   -0.090156 -0.086484 -0.041116 -0.047891   -0.051075 -0.053623   
1    0.559983  0.486570  0.255760  0.293548    0.291495  0.274975   
2   -0.332217  2.477795 -0.242874 -0.217928   -0.223670  1.674334   
3   -0.099724  0.062463 -0.100274 -0.016484   -0.065851  0.037861   
4    0.158038  0.256140  0.106968  0.076224    0.203917  0.271432   

   cheerleading  baseball    tennis    sports  ...      blonde      mall  \
0     -0.059512 -0.052457 -0.029797 -0.085470  ...   -0.026074 -0.124448   
1      0.297316  0.347514  0.151439  0.535451  ...    0.108155  0.659365   
2     -0.207327 -0.201131 -0.168939 -0.297123  ...  168.307291 -0.369915   
3     -0.113975 -0.115448  0.039235 -0.110850  ...   -0.010873 -0.084357   
4      0.440060  0.038989  0.145669  0.141436  ...    0.074666  0.659892   

   shopping   clothes  hollister  abercrombie       die      death      drunk  \
0 -0.123985 -0.158234  -0.

# Этап 5.


In [26]:
full_data_columns = data.columns.difference(['gradyear', 'gender', 'age'])
full_data = data[full_data_columns]


scaler_full = StandardScaler()
full_data_scaled = scaler_full.fit_transform(full_data)


kmeans_full = KMeans(n_clusters=5, random_state=42, n_init='auto')
kmeans_full.fit(full_data_scaled)

centroids_full = kmeans_full.cluster_centers_
centroids_full_df = pd.DataFrame(centroids_full, columns=full_data_columns)

first_ten_students_clusters_full = kmeans_full.labels_[:10]

# Вывод результатов
centroids_full_df, first_ten_students_clusters_full

(   abercrombie      band  baseball  basketball     bible      blonde  \
 0    -0.117113 -0.058046 -0.040283   -0.084441 -0.038343   -0.026686   
 1     0.022059 -0.027054  0.059547    0.297421  0.039093    0.032814   
 2    -0.183032  0.626046 -0.201131   -0.332217 -0.104247  168.307291   
 3     0.680166  0.386003  0.223100    0.451005  0.233021    0.106433   
 4     0.360657 -0.049413  0.128310    0.176163  0.027635    0.079240   
 
    cheerleading    church   clothes      cute  ...      rock        sex  \
 0     -0.146851 -0.082626 -0.159010 -0.125823  ... -0.104245  -0.099110   
 1     -0.095854  0.113307 -0.026448  0.042953  ...  0.121235  -0.036850   
 2     -0.207327 -0.297557  1.801613  0.843856  ...  1.050397  19.395543   
 3     -0.025711  0.480842  0.996792  0.745555  ...  0.624647   0.632108   
 4      5.393580  0.145263  0.145630  0.284474  ...  0.105876   0.027287   
 
        sexy  shopping    soccer  softball    sports  swimming    tennis  \
 0 -0.068424 -0.137648 -0.

# Вопросы для подготовки к лабораторной работе

### 1. В каких библиотеках Python реализован метод k-средних?

Метод k-средних реализован в следующих библиотеках Python:

- **Scikit-learn** (`sklearn.cluster.KMeans`) — одна из самых популярных библиотек для машинного обучения с поддержкой кластеризации.
- **MLlib (Spark)** — библиотека для обработки больших данных в Spark.

---

### 2. Каким образом производится обучение модели? Какие параметры необходимо указать для запуска обучения? Как проверить эффективность модели?

#### Обучение модели:
Метод k-средних начинается с задания количества кластеров `k` и случайной инициализации центроидов. Алгоритм затем итеративно перемещает центроиды, пока не достигнет сходимости, то есть до тех пор, пока точки внутри кластеров не перестанут изменять свои принадлежности.

#### Параметры для запуска модели:
- `n_clusters` — количество кластеров.
- `init` — метод инициализации центроидов (например, `'k-means++'`).
- `max_iter` — максимальное количество итераций для алгоритма.
- `random_state` — фиксирует случайность для воспроизводимости.
- `n_init` — количество запусков алгоритма с различной инициализацией центроидов.

---

### 3. Что необходимо сделать, чтобы повысить эффективность модели?
Корректный выбор количества кластеров (k):
Используйте методы, такие как метод локтя или силуэтный коэффициент, для определения оптимального числа кластеров.

Предварительная обработка данных:
Нормализация или стандартизация данных поможет избежать влияния признаков с разными масштабами на результаты кластеризации.

Инициализация центроидов:
Использование метода k-means++ позволяет выбрать более осмысленные стартовые центроиды, что может повысить точность и скорость сходимости модели.

Увеличение числа запусков (n_init):
Увеличьте количество запусков алгоритма с различной инициализацией центроидов для улучшения результатов и нахождения более оптимального решения.
